From f2592f02467e30911ef143df1a73806a9cdc7fd1 Mon Sep 17 00:00:00 2001
From: jovan <git.keseljevicjovan@proton.me>
Date: Tue, 23 Jul 2024 01:10:49 +0200
Subject: [PATCH] remove borders from nonfloating clients in monocle layoutt

---
 dwm.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/dwm.c b/dwm.c
index 67c6b2b..5416cd7 100644
--- a/dwm.c
+++ b/dwm.c
@@ -317,6 +317,10 @@ applysizehints(Client *c, int *x, int *y, int *w, int *h, int interact)
 	int baseismin;
 	Monitor *m = c->mon;
 
+	// return 1 if layout is monocle
+	if (&monocle == c->mon->lt[c->mon->sellt]->arrange)
+		return 1;
+
 	/* set minimum possible */
 	*w = MAX(1, *w);
 	*h = MAX(1, *h);
@@ -1282,6 +1286,11 @@ resizeclient(Client *c, int x, int y, int w, int h)
 	c->oldw = c->w; c->w = wc.width = w;
 	c->oldh = c->h; c->h = wc.height = h;
 	wc.border_width = c->bw;
+	if ((&monocle == c->mon->lt[c->mon->sellt]->arrange) && (!c->isfloating)) {
+		wc.border_width = 0;
+		c->w = wc.width += c->bw * 2;
+		c->h = wc.height += c->bw * 2;
+	}
 	XConfigureWindow(dpy, c->win, CWX|CWY|CWWidth|CWHeight|CWBorderWidth, &wc);
 	configure(c);
 	XSync(dpy, False);
-- 
2.45.2
